### PR TITLE
core: add insert_op_at_location to PatternRewriter

### PR DESCRIPTION
**Pull request by superlopuh**
I'm messing about with some helpers that pass around what are effectively insertion points, this functionality makes that code simpler. It also feels like a general entropy minimising step.

### CODE REVIEW

**Review by PapyChacal — APPROVED** [review]

**Review by compor — APPROVED** [review]

**Review by math-fehr — APPROVED** [review]
Very cool!

File: xdsl/pattern_rewriter.py [hunk 1]
```diff
@@ -92,60 +92,47 @@ class PatternRewriter(PatternRewriterListener):
...
+    def insert_op_at_location(
+        self, op: Operation | Sequence[Operation], insertion_point: InsertPoint
+    ):
+        """Insert operations at a certain location in a block."""
+        self.has_done_action = True
+        op = [op] if isinstance(op, Operation) else op
+        if not op:
+            return
+        for op_ in op:
+            Builder(insertion_point).insert(op_)
```
Comment (AntonLydike) [inline xdsl/pattern_rewriter.py:104]: Can we hoist the builder construction out of the loop? I'm not deeply familiar with it's API, but this seems a bit wasteful imo.

Comment (superlopuh) [reply xdsl/pattern_rewriter.py:104]: Since putting this up I've realised this is not the right approach, we should work with the rewriter directly

**Comment by superlopuh** [comment]
Updated to add equivalent method to Rewriter, seems a bit cleaner now

File: xdsl/pattern_rewriter.py [hunk 2]
```diff
@@ -92,60 +92,47 @@ class PatternRewriter(PatternRewriterListener):
...
+
+        for op_ in op:
+            self.handle_operation_insertion(op_)
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/pattern_rewriter.py]: No longer relevant :)

File: xdsl/pattern_rewriter.py [hunk 3]
```diff
@@ -86,60 +86,47 @@ class PatternRewriter(PatternRewriterListener):
...
+    def insert_op_at_location(
+        self, op: Operation | Sequence[Operation], insertion_point: InsertPoint
+    ):
+        """Insert operations at a certain location in a block."""
+        self.has_done_action = True
+        op = (op,) if isinstance(op, Operation) else op
+        if not op:
+            return
+        for op_ in op:
+            Rewriter.insert_op_at_location(op_, insertion_point)
```
Comment (math-fehr) [inline xdsl/pattern_rewriter.py:98]: Actually, `Rewriter.insert_op_at_location` should allow a sequence of operations? Or is that not the case?

File: xdsl/rewriter.py [hunk 4]
```diff
@@ -252,19 +252,23 @@ def insert_block_before(block: Block | list[Block], target: Block):
         pos = region.get_block_index(target)
         region.insert_block(block_list, pos)
 
+    @staticmethod
+    def insert_op_at_location(op: Operation, insertion_point: InsertPoint):
```
Comment (math-fehr) [inline xdsl/rewriter.py:256]: Yeah, I would add `Sequence[Operation]` as well to make things simpler

**Pull request merged**